### PR TITLE
Add SWAG Vision Transformer Weight

### PR DESCRIPTION
**Pull request by YosuaMichael**
subtask of https://github.com/pytorch/vision/issues/5708

Adding weight of ViT_B_16 and ViT_L_16 model from https://github.com/facebookresearch/SWAG

### CODE REVIEW

File: torchvision/models/vision_transformer.py [hunk 1]
```diff
@@ -284,7 +285,15 @@ def _vision_transformer(
...
-    image_size = kwargs.pop("image_size", 224)
+
+    image_size = None
+    if "image_size" in kwargs:
+        image_size = kwargs.pop("image_size", None)
+    if image_size is None and weights is not None and "size" in weights.meta:
+        image_size = weights.meta["size"]
+        if isinstance(image_size, abc.Sequence):
+            image_size = image_size[0]
+    image_size = image_size or 224
```
Comment (YosuaMichael) [inline torchvision/models/vision_transformer.py:296]: The weights for SWAG vit_b_16 tied to `image_size=384`. I edited this part so the users can just call:
```
m = vit_b_16(ViT_B_16_Weights.IMAGENET1K_SWAG_V1)
```
instead of
```
m = vit_b_16(ViT_B_16_Weights.IMAGENET1K_SWAG_V1, image_size=384)
```
However, there might be a better way to handle this. Any suggestions are welcome :)

Comment (datumbox) [reply torchvision/models/vision_transformer.py:296]: That's a good approach. I would only refactor slightly to simplify (use if/elif/else). Other than that, your approach looks good.

File: torchvision/models/vision_transformer.py [hunk 2]
```diff
@@ -313,6 +322,15 @@ def _vision_transformer(
     "interpolation": InterpolationMode.BILINEAR,
 }
 
+_COMMON_SWAG_META = {
+    "task": "image_classification",
+    "architecture": "ViT",
+    "publication_year": 2022,
+    "recipe": "https://github.com/facebookresearch/SWAG",
+    "categories": _IMAGENET_CATEGORIES,
+    "interpolation": InterpolationMode.BICUBIC,
```
Comment (datumbox) [inline torchvision/models/vision_transformer.py:331]: Nit: Should we share some of the parameters from common and overwrite them here?

Something like:
```python
_COMMON_SWAG_META = {
    **_COMMON_META,
    "publication_year": 2022,
    "recipe": "https://github.com/facebookresearch/SWAG",
 }
```

We also need to confirm what the interpolation that they used is.

File: torchvision/models/vision_transformer.py [hunk 3]
```diff
@@ -328,6 +346,20 @@ class ViT_B_16_Weights(WeightsEnum):
             "acc@5": 95.318,
         },
     )
+    IMAGENET1K_SWAG_V1 = Weights(
+        url="https://download.pytorch.org/models/vit_b_16_swag-9ac1b537.pth",
+        transforms=partial(
+            ImageClassification, resize_size=384, interpolation=InterpolationMode.BICUBIC, crop_size=384
```
Comment (datumbox) [inline torchvision/models/vision_transformer.py:352]: We need to confirm this config with the team.

File: torchvision/models/vision_transformer.py [hunk 4]
```diff
@@ -328,6 +346,20 @@ class ViT_B_16_Weights(WeightsEnum):
...
+        ),
+        meta={
+            **_COMMON_SWAG_META,
+            "num_params": 86859496,
+            "size": (384, 384),
+            "min_size": (384, 384),
+            "acc@1": 85.29,
+            "acc@5": 97.65,
```
Comment (datumbox) [inline torchvision/models/vision_transformer.py:360]: We should put here the stats from a `total_batch_size=1` run. We usually do 3 digit accuracy.

File: torchvision/models/vision_transformer.py [hunk 5]
```diff
@@ -328,6 +346,20 @@ class ViT_B_16_Weights(WeightsEnum):
...
         },
     )
+    IMAGENET1K_SWAG_V1 = Weights(
+        url="https://download.pytorch.org/models/vit_b_16_swag-9ac1b537.pth",
+        transforms=partial(
+            ImageClassification, resize_size=384, interpolation=InterpolationMode.BICUBIC, crop_size=384
+        ),
+        meta={
+            **_COMMON_SWAG_META,
+            "num_params": 86859496,
```
Comment (datumbox) [inline torchvision/models/vision_transformer.py:356]: I assume this is different from the one at `IMAGENET1K_V1` because of the size of the input. Our tests currently make the assumption that all weights have the same exactly number of parameters, this is true for 99% of the models (actually this one is the first that has this issue). So we will need to update the test logic to account for deliberate differences like this. What do you propose?

**Review by datumbox — COMMENTED** [review]
Thanks @YosuaMichael. I've added a couple of comments but overall I like your approach:

File: torchvision/models/vision_transformer.py (hunk 2 above)
Comment (YosuaMichael) [reply torchvision/models/vision_transformer.py:331]: oh, didn't know we can overwrite like that! Good point!

For the interpolation, I got it from their eval method: https://github.com/facebookresearch/SWAG/blob/main/imagenet_1k_eval.py#L198

and to me it seems like they are using `BICUBIC`.

File: torchvision/models/vision_transformer.py (hunk 5 above)
Comment (YosuaMichael) [reply torchvision/models/vision_transformer.py:356]: I think we can do as you said in the chat:

> The specific line on the test is written this way to avoid slowdowns related to downloading and loading the model. I think it's a good trick for 99% of the models but when it fails we should check explicitly if the number of parameters match the meta-data. Aka do the same as line: https://github.com/pytorch/vision/blob/main/test/test_extended_models.py#L115

So if the fast check fails, a more thorough check is done prior marking it as incorrect. What you think?

will add a commit for this

File: torchvision/models/vision_transformer.py [hunk 6]
```diff
@@ -284,10 +285,20 @@ def _vision_transformer(
...
 ) -> VisionTransformer:
-    image_size = kwargs.pop("image_size", 224)
-
     if weights is not None:
         _ovewrite_named_param(kwargs, "num_classes", len(weights.meta["categories"]))
+        if "size" in weights.meta:
+            if isinstance(weights.meta["size"], int):
+                _ovewrite_named_param(kwargs, "image_size", weights.meta["size"])
+            elif isinstance(weights.meta["size"], abc.Sequence):
+                torch._assert(
```
Comment (datumbox) [inline torchvision/models/vision_transformer.py:294]: FYI: Here you can use Exceptions because it executed during the construction of the model. The issue is when Exceptions appear in the forward. Up to you if you want to keep it as is.

File: torchvision/models/vision_transformer.py [hunk 7]
```diff
@@ -313,6 +324,12 @@ def _vision_transformer(
     "interpolation": InterpolationMode.BILINEAR,
 }
 
+_COMMON_SWAG_META = {
+    **COMMON_META,
+    "recipe": "https://github.com/facebookresearch/SWAG",
```
Comment (datumbox) [inline torchvision/models/vision_transformer.py:329]: "publication_year: 2022,

**Review by datumbox — COMMENTED** [review]
Looking good. Could you provide also the commands that you used to verify the model along with their output? Something similar to https://github.com/pytorch/vision/pull/5450#issuecomment-1053560434 (note that the commands will be slightly different because the Multi-weight support API is out of prototype).

File: README.rst [hunk 8]
```diff
@@ -186,9 +186,9 @@ This is a utility library that downloads and prepares public datasets. We do not
 
 If you're a dataset owner and wish to update any part of it (description, citation, etc.), or do not want your dataset to be included in this library, please get in touch through a GitHub issue. Thanks for your contribution to the ML community!
 
-Disclaimer on Models
-====================
+Pre-trained Model License
+=========================
 
-Pretrained models provided in this library may have their own license or term and condition to use. It is your responsibility to determine whether you have permission to use the model for your use case.
+Pre-trained models provided in this library may have their own license or terms and conditions to use. It is your responsibility to determine whether you have permission to use the model for your use case.
```
[the remark this answers is not in the document]
Comment (YosuaMichael) [reply README.rst:192]: sure, will add the command here

**Comment by YosuaMichael** [comment]
@datumbox here is the command I used for testing the `vit_b_16_swag` model:
```
python -u ~/script/run_with_submitit.py --timeout 3000 --ngpus 1 --nodes 1 --partition train --model vit_l_16 --data-path="/datasets01_ontap/imagenet_full_size/061417" --test-only --batch-size=1 --weights="ViT_L_16_Weights.IMAGENET1K_SWAG_V1"
```

**Review by datumbox — APPROVED** [review]
Thanks a lot of the changes @YosuaMichael, everything looks good to me.

@mannatsingh Could you have also a look to let us know if all look good to you?

**Review by mannatsingh — APPROVED** [review]
Everything looks perfect, thanks @YosuaMichael !

**Comment by datumbox** [comment]
The [failures](https://app.circleci.com/pipelines/github/pytorch/vision/16280/workflows/9a63e50e-845b-4c9a-b6c4-29df878ed591/jobs/1319486) are unrelated. cc @pmeier 

@YosuaMichael you are good to merge.
